### PR TITLE
allow to create multiple link shares via share api

### LINKED ISSUE #10434: Share link not visible because of ad blockers
### Steps to reproduce
1. Click on share symbol for file or folder
2. Check out side panel for sharing
3. Miss share link option

### Expected behaviour
I expect an option to share a link.

### Actual behaviour
Only input for user sharing is visible
capture

### Server configuration

**Nextcloud version:** 14.0.0 Beta 1

**Updated from an older Nextcloud/ownCloud or fresh install:** fresh

**Where did you install Nextcloud from:**

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Firefox 61.0.1 (64-bit)

**Operating system:** Win 10

#### Browser log
<details>
<summary>Browser log</summary>

```
Content Security Policy: Directive ‘child-src’ has been deprecated. Please use directive ‘worker-src’ to control workers, or directive ‘frame-src’ to control frames respectively.
JQMIGRATE: Migrate is installed, version 1.4.0 core.js:7:542
Content Security Policy: The page’s settings blocked the loading of a resource at https://cloud.example.com/index.php/apps/files/ (“script-src”).
<script> source URI is not allowed in this document: “https://cloud.example.com/apps/notifications/js/notifications.js?v=1f8ad6d7-0”. files:51
unreachable code after return statement[Learn More] merged.js:1494:4
window.controllers/Controllers is deprecated. Do not use it for UA detection. merged.js:2171
New search handler registered search.js:66:4
querying Whats New data was successful: nocontent merged-template-prepend.js:4455:4
undefined  merged-template-prepend.js:4456:4
```
</details>

### LINKED ISSUE #11169: Allow multiple share links
Before starting on the implementation of multiple share links I would like to get some feedback from @nextcloud/designers  how this should look like.

I already thought about it a little bit and this is my suggestion:

1. Initial share view:

If you click on "create new public link" you get a text input field which is pre-filled with "share link 1", the next one "share link 2", and so on, so that the user can give the link a name.

If the link was created it will look like this, the user sees the new share link, with the '...' menu and again a option to create another public link:

To the '...' menu I would add the two new additional option "rename" and "unshare".

Opionions? cc @jancborchardt

# Steps to complete it

- [ ] agree on design
- [ ] front-end implementation
- [ ] back-end implementation
- [ ] OCS API implementation
- [ ] Tests

### PR DESCRIPTION
**Pull request by schiessle**
- [x] Allow to create multiple share links (API level)
- [x] Get multiple share links (API level)
- [x] allow to give share links unique names (API level)
- [x] allow to change unique name of a share (API level)
- [x] Allow to create multiple share links (web ui)
- [x] Show multiple share links (web ui)
- [ ] allow to give share links unique names (web ui)
- [ ] allow to change unique name of a share (web ui)

for ux discussion, etc see  https://github.com/nextcloud/server/issues/11169
Fix #10434
Closes #11537 : not possible, we cannot wait for the request of a new share to be done as the `.done` or `success` hooks generate a new scope and a clipboard triggered by a click needs to be in a direct call of a function triggered by an event :(
Fix #11169

### CODE REVIEW

**Comment by schiessle** [comment]
cc @AndyScherzinger  @tobiasKaminsky @marinofaggiana @camilasan @rullzer  With Nextcloud 15 we will support multiple share links. This PR has already implement everything you need to start implementing it on the client. Just in case you looking for something to do :wink: 

Creating links works the same as before. There is just one additional parameter you can use "label" to create/update link shares so that the user can give the share link names to remember which link was used for what.

**Comment by marinofaggiana** [comment]
ToDo added :-)

File: core/Migrations/Version13000Date20170718121200.php [hunk 1]
```diff
@@ -397,6 +397,10 @@ public function changeSchema(IOutput $output, \Closure $schemaClosure, array $op
 				'notnull' => false,
 				'length' => 64,
 			]);
+			$table->addColumn('label', 'string', [
```
Comment (skjnldsv) [inline core/Migrations/Version13000Date20170718121200.php]: Hum, shouldn't we create a newer file for that?

**Comment by skjnldsv** [comment]
Should be finished today!
@jancborchardt @nextcloud/designers Can we try to find another saying for "Share link" I'm always finding this confusing as it looks like an action and not a statement :)


Can you start a review now?
That would increase my response time to fixes to day for the feature freeze :)

Basically the only thing left is the label handling! How should we do that?
1. Add an edit button?
2. In the popovermenu?

**Comment by jancborchardt** [comment]
Just beginning to test this! :) First feedback:
- The ➕ icon is not keyboard-accessible, no highlight on focus
- Ideally the whole row should be clickable, not just the ➕
- Not sure it’s screenreader-accessible right now with the "title" on the icon. But it would be when the whole row is the action.

**Comment by jancborchardt** [comment]
Made the background of the share link primary color instead of bland grey. This is in line with the change in the Talk app at https://github.com/nextcloud/spreed/pull/1274

**Comment by skjnldsv** [comment]
> Made the background of the share link primary color instead of bland grey. This is in line with the change in the Talk app at [nextcloud/spreed#1274](https://github.com/nextcloud/spreed/pull/1274)

I was afraid it would be too much, but it is spot on!! :+1:

**Comment by skjnldsv** [comment]
@nextcloud/designers please review!
What to test:
1. public sharing
2. with expire date enforced
3. with password open by default
4. with password enforced
5. with default expire date 
6. note management

It is still missing the label, but I rather have early feedback to fix things asap!

**Comment by skjnldsv** [comment]
> * Ideally the whole row should be clickable, not just the heavy_plus_sign

I'm afraid putting the click to the whole row will also trigger other stuff in the children (popover or such) I thing I will rather keep it on the button only to avoid creating issues :(

**Comment by rullzer** [comment]
The hide download checkbox does not work anymore

(is only there for file shares now)

**Comment by schiessle** [comment]
> Basically the only thing left is the label handling! How should we do that?
> 1. Add an edit button?
> 2. In the popovermenu?

My suggestion would be to have a "rename" action in the "..." menu. Which behaves similar to renaming a file/folder. If you click the rename action the popover closes and the label "share link" turns into a input field where the user can change the name and press enter or focus out.

File: core/js/sharedialoglinkshareview.js [hunk 2]
```diff
@@ -143,46 +158,76 @@
...
+			var $element = $(event.target);
+			var $li = $element.closest('li[data-share-id]');
+			var $el = $li.find('.linkText');
 			$el.focus();
 			$el.select();
 		},
 
 		onHideDownloadChange: function() {
-			var $checkbox = this.$('.hideDownloadCheckbox');
+			var $element = $(event.target);
```
Comment (rullzer) [inline core/js/sharedialoglinkshareview.js]: event is not defined here.

**Review by rullzer — APPROVED** [review]
Gave it a test round. Works as advertised.

Lets fix the rest later and get this in so it doesn't grow more in size :wink:

**Comment by jancborchardt** [comment]
Pushed some sorting order and wording fixes! Someone needs to build the handlebars stuff though as I am getting errors and there is no docs! (Currently stuck at `Error: Cannot find module 'optimist'`)

----------------------

> My suggestion would be to have a "rename" action in the "..." menu. Which behaves similar to renaming a file/folder. If you click the rename action the popover closes and the label "share link" turns into a input field where the user can change the name and press enter or focus out.

> Yes, I also like this, but the menu is already very filled :(

I would agree with @schiessle here to put it in the  3-dot menu but to be honest this function is just utterly useless.

 -------------------

Because I’m a bit annoyed at this shoveling of stuff into menus and sidebars yet again, I opened an issue for cleaning up that popovermenu: https://github.com/nextcloud/server/issues/12178

Bottom line: We need to stop just stupidly adding stuff! No one will be able to find their favorite custom setting among all the other custom settings. It’s nuts!

**Comment by danxuliu** [comment]
>  Someone needs to build the handlebars stuff 

Done.

Besides that I have cleaned up a little the commit history; the code is almost the same as it was before, except for a removed `console.error(e)` (because it was the only content of a commit named _Various fixes_, so I dropped it), an updated documentation to reflect that `ShareItemModel` no longer has the attributes `isLinkShare` and `link`, and a "cast" of the attribute `hideDownload` to a boolean from the int `share.hide_download`.

@skjnldsv If you already made a commit to fix the JavaScript unit tests cherry-picking it on top of the current commits should cause no conflicts.

File: core/js/shareitemmodel.js [hunk 3]
```diff
@@ -864,20 +863,12 @@
 						} else {
 							link += OC.generateUrl('/s/') + share.token;
 						}
```
Comment (danxuliu) [inline core/js/shareitemmodel.js]: All the code above to set the link is now dead code, because the `link` attribute is no longer set in the share (and the `url` attribute is now used instead of `link` when copying the link to the clipboard); I do not know what that _pre-token_ thing is, though, and if it can be safely replaced by the URL attribute returned by the server.

**Review by danxuliu — DISMISSED** [review]
JavaScript unit tests need to be fixed before merging.

Besides that, adding a new link share after the first one does not work when passwords are enforced (the menu to set the password is not shown).

File: core/js/shareitemmodel.js (hunk 3 above)
Comment (skjnldsv) [reply core/js/shareitemmodel.js]: No idea. I usually don't want to touch too much here. 😂

**Comment by skjnldsv** [comment]
> Besides that, adding a new link share after the first one does not work when passwords are enforced (the menu to set the password is not shown).

Nice catch!!

Thanks a lot for the fixes! You two are completely crazy working through the night :(
You need to sleep!

**Review by MorrisJobke — APPROVED** [review]
Tested and works nice 👍 Current state is also perfectly fine to be merged and iterated on later

**Comment by danxuliu** [comment]
@jancborchardt By the way, same problem as described in https://github.com/nextcloud/spreed/pull/1274#pullrequestreview-170646080:
share-link-icon-theming-color-white

**Review by danxuliu — CHANGES_REQUESTED** [review]
>  Fixed

Not so fast; I have some complaints ;-)

File: core/js/tests/specs/sharedialoglinkshareview.js [hunk 4]
```diff
@@ -81,12 +81,17 @@ describe('OC.Share.ShareDialogLinkShareView', function () {
...
-				linkShare: {
-					isLinkShare: true,
-				}
+				linkShares: [{
+					token: 'thetoken',
+					url: 'https://nextcloud.com',
+					expiration: '',
+					permissions: OC.PERMISSION_READ,
+					stime: 1403884258,
+					id: 123
```
Comment (danxuliu) [inline core/js/tests/specs/sharedialoglinkshareview.js]: Are all the parameters above needed for the tests? Or is it enough to just give the id?

File: core/js/tests/specs/sharedialoglinkshareview.js [hunk 5]
```diff
@@ -141,16 +151,22 @@ describe('OC.Share.ShareDialogLinkShareView', function () {
...
+					expiration: '',
+					permissions: OC.PERMISSION_READ,
+					stime: 1403884258,
+					id: 321,
+					hide_download: true
+				}]
 			});
+			view.render();
 
 			$hideDownloadCheckbox = view.$el.find('.hideDownloadCheckbox');
```
Comment (danxuliu) [inline core/js/tests/specs/sharedialoglinkshareview.js]: No `data-share-id` used here even if the id `321` was used in `shareModel.set` above; maybe it is unneeded in the other test too?

File: core/js/tests/specs/sharedialogviewSpec.js [hunk 6]
```diff
@@ -165,47 +177,57 @@ describe('OC.Share.ShareDialogView', function() {
...
-				token: 'tehtoken',
-				link: link,
-				expiration: '',
-				permissions: OC.PERMISSION_READ,
-				stime: 1403884258,
+				OC.generateUrl('/s/') + 'thetoken';
+			shareModel.set({
+				linkShares: [{
+					token: 'thetoken',
+					url: link,
```
Comment (danxuliu) [inline core/js/tests/specs/sharedialogviewSpec.js]: The link is now got from the share returned by the server, so it can be hardcoded like the other parameters; no need to generate it above.

**Comment by skjnldsv** [comment]
@danxuliu be quick!
I want to get this in today! And don't use this as an excuse to merge your pr before mine :p

**Review by danxuliu — DISMISSED** [review]

File: core/js/sharedialoglinkshareview.js [hunk 7]
```diff
@@ -537,7 +537,7 @@
 
 			var minPasswordLength = 4
 			// password policy?
-			if(oc_capabilities.password_policy && oc_capabilities.password_policy.minLength) {
+			if(window.oc_capabilities && oc_capabilities.password_policy && oc_capabilities.password_policy.minLength) {
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply core/js/sharedialoglinkshareview.js]: Because it seems there is an issue between the declaration of variables and window.xxx :)

File: core/js/tests/specs/sharedialoglinkshareview.js [hunk 8]
```diff
@@ -107,7 +112,6 @@ describe('OC.Share.ShareDialogLinkShareView', function () {
 		it('is not shown if the share is a folder', function() {
 			shareModel.fileInfoModel.set('mimetype', 'httpd/unix-directory');
 
-			// Setting the item type also triggers the rendering
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply core/js/tests/specs/sharedialoglinkshareview.js]: No, I changed it because it was unnecessary.
This is done on purpose.

File: core/js/tests/specs/sharedialoglinkshareview.js [hunk 9]
```diff
@@ -141,16 +151,22 @@ describe('OC.Share.ShareDialogLinkShareView', function () {
 			$hideDownloadCheckbox.change();
 
 			expect($workingIcon.hasClass('hidden')).toBeFalsy();
-			expect(shareModel.saveLinkShare.withArgs({ hideDownload: true }).calledOnce).toBeTruthy();
+			expect(shareModel.saveLinkShare.calledOnce).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply core/js/tests/specs/sharedialoglinkshareview.js]: None, forgot to edit ;)

File: core/js/tests/specs/sharedialoglinkshareview.js (hunk 4 above)
Comment (skjnldsv) [reply core/js/tests/specs/sharedialoglinkshareview.js]: No idea if they're needed. The id is at least yes.

File: core/js/tests/specs/sharedialoglinkshareview.js (hunk 5 above)
Comment (skjnldsv) [reply core/js/tests/specs/sharedialoglinkshareview.js]: It is! :)
There is only one share, so it doesn't really matter ;)

**Review by danxuliu — CHANGES_REQUESTED** [review]
>  Fixed :D

Almost :-P I will send a final fix.

**Comment by skjnldsv** [comment]
Failure unrelated: `sh: 1: kill: No such process` https://drone.nextcloud.com/nextcloud/server/12248/283

**Comment by skjnldsv** [comment]
> Almost :-P I will send a final fix.

Ahaha! Okay :)
Thanks a lot for your help @danxuliu :hugs:

File: core/js/sharedialoglinkshareview.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (danxuliu) [reply core/js/sharedialoglinkshareview.js]: OK, but the regular code is not the place to address that ;-) Values set by [JSConfigHelper](https://github.com/nextcloud/server/blob/master/lib/private/Template/JSConfigHelper.php#L161) should be set in the JavaScript unit tests in [specHelper.js](https://github.com/nextcloud/server/blob/4a0b7aaf6c80f59552b1da861432476a8ac4c7cf/core/js/tests/specHelper.js#L86)

File: core/js/tests/specs/sharedialoglinkshareview.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (danxuliu) [reply core/js/tests/specs/sharedialoglinkshareview.js]: Ha, I can see through your lies! :-P [Setting the item type still triggers the rendering](https://github.com/nextcloud/server/pull/11844/files#diff-da7df34db49581af77bd59e4f269dc36R88) ;-)

Setting the link share does no longer trigger the rendering, though; although you did not that on purpose, it was a side effect of renaming the `linkShare` attribute to `linkShares` in the model :-P (I also [removed the event handler](https://github.com/nextcloud/server/pull/11844/files#diff-da7df34db49581af77bd59e4f269dc36L92) because it was no longer used).

File: core/js/tests/specs/sharedialoglinkshareview.js (hunk 4 above)
Comment (danxuliu) [reply core/js/tests/specs/sharedialoglinkshareview.js]: No, they were not, so I removed them everywhere ;-)

**Review by danxuliu — APPROVED** [review]
Rejoice; finally, the nitpicker is _reasonably_ satisfied :+1:

I hope Drone is too... :-P

**Comment by MorrisJobke** [comment]
🎉 I will merge once drone has done it's job 🎉

File: core/js/sharedialoglinkshareview.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (skjnldsv) [reply core/js/sharedialoglinkshareview.js]: You learn something everyday!
Well, calling the variable or the window.xxx is the same when declared as global :)

[the remark this answers is not in the document]
Comment (skjnldsv) [reply core/js/sharedialoglinkshareview.js]: Done

**Comment by lopezio** [comment]
> Future interface stuff that is optional:
> > * [ ]  allow to give share links unique names (web ui)
> > * [ ]  allow to change unique name of a share (web ui)

@schiessle @jancborchardt any progress on these points..? I think they're fundamental to make the multiple linking functionality usable..  with some kind of distinction between the shared links it would be possible to explicitely delete or re-share links based on a target public / individual..

Best Regards!
